src/datastore/pg-write-store.ts

```diff
@@ -1682,15 +1681,33 @@ export class PgWriteStore extends PgStore {
         tenure_change_signature: tx.tenure_change_signature ?? null,
         tenure_change_signers: tx.tenure_change_signers ?? null,
       }));
+      // The incoming mempool transactions might have already been settled
+      // We need to mark them as pruned to avoid inconsistent tx state
       const result = await sql<{ tx_id: string }[]>`
         WITH inserted AS (
           INSERT INTO mempool_txs ${sql(values)}
           ON CONFLICT ON CONSTRAINT unique_tx_id DO NOTHING
           RETURNING tx_id
         ),
+        settled AS (
+          SELECT tx_id
+          FROM txs
+          WHERE
+            tx_id IN ${sql(values.map(b => b.tx_id))} AND
+            canonical = true AND
+            microblock_canonical = true
+        ),
+        pruned AS (
+          UPDATE mempool_txs
```

Comment:
This query is looking great! However, can you add a test for this scenario @bestmike007 ? I'm worried that a row that is both inserted and updated within the same query will trigger a "cannot affect row twice" error in Postgres.

If that is the case you'll be forced to split the query in two stages. A test will help us determine that.

Reply:
It's worse, it succeeds with unexpected behavior. Let's revert it to using 2 stages.

![image](https://github.com/hirosystems/stacks-blockchain-api/assets/432106/32c051a0-103a-40d1-aefa-df9a6cf47cc8)

Reply:
Reverted.